### ISSUE TITLE
Auto-Configuration Error: Cannot find cudnn.h at /usr/lib/x86_64-linux-gnu/include/cudnn.h

### ISSUE THREAD

**Issue by byzhang** [tensorflow/tensorflow#8264]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
#4397, but it's for 16.04

### Environment info
Operating System: Ubuntu 14.04
Installed version of CUDA and cuDNN: 8.0, 5.1
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
```
ls -l /path/to/cuda/lib/libcud*
-rw-r--r-- 1 root 543K Jan 26 15:48 /usr/local/cuda/lib64/libcudadevrt.a
lrwxrwxrwx 1 root   16 Jan 26 15:51 /usr/local/cuda/lib64/libcudart.so -> libcudart.so.8.0
lrwxrwxrwx 1 root   19 Jan 26 15:51 /usr/local/cuda/lib64/libcudart.so.8.0 -> libcudart.so.8.0.61
-rw-r--r-- 1 root 406K Jan 26 15:48 /usr/local/cuda/lib64/libcudart.so.8.0.61
-rw-r--r-- 1 root 757K Jan 26 15:48 /usr/local/cuda/lib64/libcudart_static.a
```
```
ll /usr/lib/x86_64-linux-gnu/libcudnn*                                                                                                                                                                
lrwxrwxrwx 1 root  29 Mar  1 11:58 /usr/lib/x86_64-linux-gnu/libcudnn.so -> /etc/alternatives/libcudnn_so
lrwxrwxrwx 1 root  18 Nov  6 23:19 /usr/lib/x86_64-linux-gnu/libcudnn.so.5 -> libcudnn.so.5.1.10
-rw-r--r-- 1 root 81M Nov  6 23:19 /usr/lib/x86_64-linux-gnu/libcudnn.so.5.1.10
lrwxrwxrwx 1 root  32 Mar  1 11:58 /usr/lib/x86_64-linux-gnu/libcudnn_static.a -> /etc/alternatives/libcudnn_stlib
-rw-r--r-- 1 root 68M Nov  6 23:19 /usr/lib/x86_64-linux-gnu/libcudnn_static_v5.a
```

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`): e895d5ca395c2362df4f5c8f08b68501b41f8a98
2. The output of `bazel version`: 
```
INFO: Reading 'startup' options from /root/.bazelrc: --batch
Build label: 0.4.2
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Dec 7 18:47:11 2016 (1481136431)
Build timestamp: 1481136431
Build timestamp as int: 1481136431
```

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
I ran
```
./configure
```
It failed with
```
INFO: Reading 'startup' options from /root/.bazelrc: --batch
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.
INFO: Reading 'startup' options from /root/.bazelrc: --batch
ERROR: package contains errors: tensorflow/compiler/tests.
ERROR: error loading package 'tensorflow/compiler/tests': Encountered error while reading extension file 'cuda/build_defs.bzl': no such package '@local_config_cuda//cuda': Traceback (most recent call last):
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 816
                _create_cuda_repository(repository_ctx)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 729, in _create_cuda_repository
                _get_cuda_config(repository_ctx)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 585, in _get_cuda_config
                _cudnn_version(repository_ctx, cudnn_install_base..., ...)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 296, in _cudnn_version
                _find_cuda_define(repository_ctx, cudnn_install_base..., ...)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 271, in _find_cuda_define
                auto_configure_fail("Cannot find cudnn.h at %s" % st...))
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 93, in auto_configure_fail
                fail("
%sAuto-Configuration Error:%s ...))

Auto-Configuration Error: Cannot find cudnn.h at /usr/lib/x86_64-linux-gnu/include/cudnn.h
```

I noticed the cudnn.h is in /usr/include/cudnn.h somehow, while its libs are in /usr/lib/x86_64-linux-gnu/.

I read cuda_configure.bzl, but didn't find an easy way to let it find both cudnn headers and libs.

**Comment by prb12** [comment]
I am unable to reproduce this error with the information you provided.
When I sync to e895d5c, the configure script works fine for me.  
At what point is it failing?  the error messages seem to be coming from bazel?

Could you please capture what responses you provided to the ./configure script and all the command lines you typed.

Also, could you please check that you installed all required dependencies as described in this page:
https://www.tensorflow.org/install/install_sources

**Comment by byzhang** [comment]
sure.
```
 ❯ ./configure                                                                                                                                                                                           Please specify the location of python. [Default is /usr/bin/python]:
Please specify optimization flags to use during compilation [Default is -march=native]:
Do you wish to use jemalloc as the malloc implementation? (Linux only) [Y/n]
jemalloc enabled on Linux
Do you wish to build TensorFlow with Google Cloud Platform support? [y/N]
No Google Cloud Platform support will be enabled for TensorFlow
Do you wish to build TensorFlow with Hadoop File System support? [y/N]
No Hadoop File System support will be enabled for TensorFlow
Do you wish to build TensorFlow with the XLA just-in-time compiler (experimental)? [y/N]
No XLA support will be enabled for TensorFlow
Found possible Python library paths:
  /usr/local/lib/python2.7/dist-packages
  /usr/lib/python2.7/dist-packages
Please input the desired Python library path to use.  Default is [/usr/local/lib/python2.7/dist-packages]

Using python library path: /usr/local/lib/python2.7/dist-packages
Do you wish to build TensorFlow with OpenCL support? [y/N]
No OpenCL support will be enabled for TensorFlow
Do you wish to build TensorFlow with CUDA support? [y/N] y
CUDA support will be enabled for TensorFlow
Please specify which gcc should be used by nvcc as the host compiler. [Default is /usr/bin/gcc]:
Please specify the CUDA SDK version you want to use, e.g. 7.0. [Leave empty to use system default]: 8.0
Please specify the location where CUDA 8.0 toolkit is installed. Refer to README.md for more details. [Default is /usr/local/cuda]:
Please specify the Cudnn version you want to use. [Leave empty to use system default]: 5
Please specify the location where cuDNN 5 library is installed. Refer to README.md for more details. [Default is /usr/local/cuda]: /usr/lib/x86_64-linux-gnu
Please specify a list of comma-separated Cuda compute capabilities you want to build with.
You can find the compute capability of your device at: https://developer.nvidia.com/cuda-gpus.
Please note that each additional compute capability significantly increases your build time and binary size.
[Default is: "3.5,5.2"]: 3.0,3.5,5.2,6.0,6.1
INFO: Reading 'startup' options from /root/.bazelrc: --batch
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.
INFO: Reading 'startup' options from /root/.bazelrc: --batch
ERROR: package contains errors: tensorflow/stream_executor.
ERROR: error loading package 'tensorflow/stream_executor': Encountered error while reading extension file 'cuda/build_defs.bzl': no such package '@local_config_cuda//cuda': Traceback (most recent call last):
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 815
                _create_cuda_repository(repository_ctx)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 728, in _create_cuda_repository
                _get_cuda_config(repository_ctx)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 584, in _get_cuda_config
                _cudnn_version(repository_ctx, cudnn_install_base..., ...)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 295, in _cudnn_version
                _find_cuda_define(repository_ctx, cudnn_install_base..., ...)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 270, in _find_cuda_define
                auto_configure_fail("Cannot find cudnn.h at %s" % st...))
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 93, in auto_configure_fail
                fail("
%sAuto-Configuration Error:%s ...))

Auto-Configuration Error: Cannot find cudnn.h at /usr/lib/x86_64-linux-gnu/include/cudnn.h
.
```

The cudnn.h was installed in
```
/usr/include/cudnn.h
```
which links to 
```
lrwxrwxrwx 1 root 26 Mar  1 11:58 /usr/include/cudnn.h -> /etc/alternatives/libcudnn
```
and again
```
lrwxrwxrwx 1 root 40 Mar  1 11:58 /etc/alternatives/libcudnn -> /usr/include/x86_64-linux-gnu/cudnn_v5.h
```
I read cuda_configure.bzl, but didn't find an easy way to let it find the cudnn headers.

**Comment by prb12** [comment]
How did you install cuDNN?  This normally results in a directory containing subdirs for `include` and `lib` so a single path serves to locate both the library and header files.

**Comment by byzhang** [comment]
It's a docker using nvidia/cuda:8.0-cudnn5-devel-ubuntu14.04

Thanks,
-B

**Comment by prb12** [comment]
Could you please take a look at the instructions [here](https://www.tensorflow.org/install/install_linux#gpu_support) for using an nvidia GPU docker container and see if there are any steps you missed?

(admittedly, that page seems to mention TF0.12 so it may be slightly out of date)

@gunan Do you have anything to add here?

**Comment by gunan** [comment]
I think this happens when cudnn is installed using the deb package. @byzhang could you confirm how you installed cudnn?

**Comment by byzhang** [comment]
I'm following
https://gitlab.com/nvidia/cuda/blob/ubuntu14.04/8.0/devel/cudnn5/Dockerfile
to install the cudnn within the docker:
```

RUN echo "deb http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1404/x86_64
/" > /etc/apt/sources.list.d/nvidia-ml.listENV CUDNN_VERSION
5.1.10LABEL com.nvidia.cudnn.version="${CUDNN_VERSION}"RUN apt-get
update && apt-get install -y --no-install-recommends \
libcudnn5=$CUDNN_VERSION-1+cuda8.0 \
libcudnn5-dev=$CUDNN_VERSION-1+cuda8.0 && \

rm -rf /var/lib/apt/lists/*
```


Thanks,
-B

**Comment by gunan** [comment]
As a very quick solution, you may create symlinks to cudnn headers in `/usr/local/cuda/include` and to all cudnn libraries in `/usr/local/cuda/lib64`. 
@flx42 Any idea about any changes in cudnn deb packages that may cause this?

**Comment by flx42** [comment]
> This normally results in a directory containing subdirs for include and lib so a single path serves to locate both the library and header files.

@prb12: that's not the case with deb packages as @gunan mentioned.

@gunan: I'm not sure what your question is. But the path `/usr/lib/x86_64-linux-gnu/include/cudnn.h` doesn't make sense. There is probably a problem when configuring TensorFlow.

**Comment by gunan** [comment]
I thought with deb packages, in the end we had symlinks for cudnn.h and library files that provide a structure as @prb12 described. It looks like that is not the case.
I think for the error message, what happened is user entered `/usr/lib/x86_64-linux-gnu/` as cudnn root directory. Our source code assumes what @prb12 described, so `/usr/lib/x86_64-linux-gnu/include/cudnn.h` is probably constructed by our code.

We will need to update our configure script to avoid this assumption.

**Comment by flx42** [comment]
`/usr/include/cudnn.h` is indeed a symlink, but that's for allowing users to switch cuDNN versions on the same machine through Debian alternatives:
```
/usr/include/cudnn.h -> /etc/alternatives/libcudnn -> /usr/include/x86_64-linux-gnu/cudnn_v5.h
```
cuDNN v4 was different though, the actual header was at `/usr/include/cudnn_v4.h`

**Comment by byzhang** [comment]
Yes, I took a quick fix by symlinks. But it's appreciated if the configure
scripts is more flexible.

Thanks,
-B

**Comment by gunan** [comment]
Sorry for the inconvenience @byzhang 
I intend to fix the script to handle installation through deb packages soon.
Until then, I just wanted to get you unblocked as quickly as possible using symlinks.

**Comment by nectario** [comment]
I get a similar issue:

This is a very frustrating process... Trying to build Java GPU support for Windows...

```
$ bazel build --config opt //tensorflow/java:tensorflow //tensorflow/java:libtensorflow_jni
ERROR: C:/development/projects/tensorflow/tensorflow/java/BUILD:142:1: error loading package 'tensorflow/java/src/main/native': Encountered error while reading extension file 'cuda/build_defs.bzl': no such package '@local_config_cuda//cuda': Traceback (most recent call last):
        File "C:/development/projects/tensorflow/third_party/gpus/cuda_configure.bzl", line 958
                _create_cuda_repository(repository_ctx)
        File "C:/development/projects/tensorflow/third_party/gpus/cuda_configure.bzl", line 846, in _create_cuda_repository
                _get_cuda_config(repository_ctx)
        File "C:/development/projects/tensorflow/third_party/gpus/cuda_configure.bzl", line 656, in _get_cuda_config
                _cudnn_install_basedir(repository_ctx)
        File "C:/development/projects/tensorflow/third_party/gpus/cuda_configure.bzl", line 211, in _cudnn_install_basedir
                auto_configure_fail("Cannot find cudnn install path....)
        File "C:/development/projects/tensorflow/third_party/gpus/cuda_configure.bzl", line 128, in auto_configure_fail
                fail("\n%sAuto-Configuration Error:%s...))

Auto-Configuration Error: Cannot find cudnn install path.
 and referenced by '//tensorflow/java:libtensorflow_jni.so'.
ERROR: Analysis of target '//tensorflow/java:tensorflow' failed; build aborted.
INFO: Elapsed time: 0.641s
```

**Comment by alyato** [comment]
hi @byzhang @gunan @flx42  i install the tensorflow by the source.
 i meet the similar error 
when i run `bazel build --config=opt --config=cuda //tensorflow/tools/pip_package:build_pip_package `
 

> Cuda Configuration Error: Cannot find cudnn.h under /usr/local/lib
 
The cuda is `7.5`  and cudNN is `5.0.5`
I install the cuDNN by

> cd ~
sudo tar xvf cudnn-7.5-linux-x64-v5.0.tgz
 cd cuda/include
sudo cp *.h /usr/local/include/
cd ../lib64
sudo cp lib* /usr/local/lib/
cd /usr/local/lib# sudo chmod +r libcudnn.so.5.0.5
sudo ln -sf libcudnn.so.5.0.5 libcudnn.so.5
sudo ln -sf libcudnn.so.5 libcudnn.so
sudo ldconfig

when i run `.configure`,the it show me 

> Do you wish to build TensorFlow with MPI support? [y/N] 
MPI support will not be enabled for TensorFlow
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
Configuration finished

Not the following by the link https://www.tensorflow.org/install/install_sources

> [Default is: "3.5,5.2"]: 3.0
Setting up Cuda include
Setting up Cuda lib
Setting up Cuda bin
Setting up Cuda nvvm
Setting up CUPTI include
Setting up CUPTI lib64
Configuration finished

Do you give me some advice? Thanks.

**Comment by gunan** [comment]
@nectario for windows please file a separate issue, that will need to be addressed separately.
@alyato, when you are installing cuDNN, the simplest way (AFAICR the way recommended in NVIDIA cuDNN website) is to just do this:
```
cd ~
sudo tar xvf cudnn-7.5-linux-x64-v5.0.tgz
sudo cp -R cuda /usr/local
export LB_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/cuda/lib
```
If you do the above, it should just work with configure script.
This issue is about the problems we see when we have cuDNN installed using the deb packages distributed by NVIDIA.

**Comment by nectario** [comment]
Thank you. I will file a separate issue. At the moment, I am uninstalling everything as I have several versions of Python installed.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mingxingtan** [comment]
Close due to inactivity for a long time.

**Comment by hongsoog** [comment]
The main reason is cuda version information is not stored in ```cudnn.h``` anymore in CUDA 11.X. In CUDA 11.X, the version information seems to stored in  new ```cudnn_version.h``` file. So many build tools such as cmake depends on ```cudnn.h``` for CUDA version information cannot guess CUDA version anymore.
